### PR TITLE
Simplify checkDependencyGroup to reduce compile time

### PR DESCRIPTION
**Pull request by NigelYiboYu**
The previous implementation of codegen checkDependencyGroup() performs
unnecessary checks. This change simplifies the original one but still
keeps the O(n^2) algorithm.

Signed-off-by: Nigel Yu <yunigel@ca.ibm.com>

### CODE REVIEW

**Comment by NigelYiboYu** [comment]
I'm currently in the process of testing this change to make sure nothing else is broken.
Benchmark results have shown an average of 4.95% reduction in codegen time. The 2-sigma range is 3.33%.

**Comment by NigelYiboYu** [comment]
@fjeremic  This is ready to review.

File: compiler/z/codegen/OMRRegisterDependency.cpp [hunk 1]
```diff
@@ -680,17 +680,14 @@ OMR::Z::RegisterDependencyConditions::bookKeepingRegisterUses(TR::Instruction *
    }
 
 uint32_t
-TR_S390RegisterDependencyGroup::genBitVectOfAssignableGPRs(TR::Instruction * currentInstruction, uint32_t          numberOfRegisters,
-   TR::CodeGenerator * cg)
+TR_S390RegisterDependencyGroup::genBitVectOfAssignableGPRs(uint32_t numberOfRegisters,
+                                                           TR::CodeGenerator * cg)
```
[the remark this answers is not in the document]
Comment (NigelYiboYu) [reply compiler/z/codegen/OMRRegisterDependency.cpp]: Yes, Having a consistent call parameter order is beneficial. Maybe we should have a work item to sweep the code base and reorder all call parameters like this?

File: compiler/z/codegen/OMRRegisterDependency.cpp [hunk 2]
```diff
@@ -680,17 +680,14 @@ OMR::Z::RegisterDependencyConditions::bookKeepingRegisterUses(TR::Instruction *
...
-   TR::RealRegister::RegNum realReg;
 
-   for (i = 0; i < numberOfRegisters; i++)
+   for (uint32_t i = 0; i < numberOfRegisters; i++)
       {
-      realReg = _dependencies[i].getRealRegister();
+      TR::RealRegister::RegNum realReg = _dependencies[i].getRealRegister();
       if (TR::RealRegister::isGPR(realReg))
          {
          availRegMap &= (~TR::RealRegister::getBitMask(realReg));
```
[the remark this answers is not in the document]
Comment (NigelYiboYu) [reply compiler/z/codegen/OMRRegisterDependency.cpp]: Done

File: compiler/z/codegen/OMRRegisterDependency.cpp [hunk 3]
```diff
@@ -700,141 +697,221 @@ TR_S390RegisterDependencyGroup::genBitVectOfAssignableGPRs(TR::Instruction * cur
...
+   if (numberOfRegisters <= 1)
       {
-      virtReg = _dependencies[i].getRegister(cg);
-      realReg = _dependencies[i].getRealRegister();
+      return;
+      }
 
-      j = i + 1;
+   bool isHighWordSupported = cg->supportsHighWordFacility() && !cg->comp()->getOption(TR_DisableHighWordRA);
+   int32_t i = 0, j = 0;
```
[the remark this answers is not in the document]
Comment (NigelYiboYu) [reply compiler/z/codegen/OMRRegisterDependency.cpp]: Done

File: compiler/z/codegen/OMRRegisterDependency.cpp [hunk 4]
```diff
@@ -700,141 +697,221 @@ TR_S390RegisterDependencyGroup::genBitVectOfAssignableGPRs(TR::Instruction * cur
...
-   cg->setAvailableHPRSpillMask(0xFFFF0000);
+/**
+*
+* \brief Handle Register Pairs
+*
+* 1. count the number of required register pairs `numReqPairs`
+* 2. count the number of available register pairs `numAvailPairs`
+* 3. Fail the compilation if `numAvailPairs` < `numReqPairs`
+*
+**/
```
[the remark this answers is not in the document]
Comment (NigelYiboYu) [reply compiler/z/codegen/OMRRegisterDependency.cpp]: * Will update the function documentation

* Will rename ```checkRegisterPairSufficiency``` to ```checkRegisterPairSufficiencyAndHPRAssignment```. This function does two things: check for register pairs within a *group* of register-dependencies AND sets HPR and 64-bit register assignment related flags.

* Will change *numberOfRegisters* to *numOfDependencies*. This is the number of pre or post-conditions within a *DependencyGroup*.

File: compiler/z/codegen/OMRRegisterDependency.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (fjeremic) [reply compiler/z/codegen/OMRRegisterDependency.cpp]: > Maybe we should have a work item to sweep the code base and reorder all call parameters like this?

Definitely! Do you want to open one up?

File: compiler/z/codegen/OMRRegisterDependency.cpp [hunk 5]
```diff
@@ -680,167 +680,269 @@ OMR::Z::RegisterDependencyConditions::bookKeepingRegisterUses(TR::Instruction *
...
-   // Count the number of pairs remaining after deps are filled
-   for (i = 0; i < NUM_S390_GPR; i += 2)
-      {
-      if ((availRegMap >> i) & 0x3 == 0x3)
-         {
-         numAvailPairs++;
-         }
-      }
+   uint32_t numReqPairs = 0, numAvailPairs = 0;
+   int32_t i;
```
Comment (fjeremic) [inline compiler/z/codegen/OMRRegisterDependency.cpp]: This declaration can be pushed into the for loop (thus reducing the scope of the variable and making a more idiomatic looking loop). Similarly elsewhere in the commit.

**Comment by charliegracie** [comment]
genie-omr build zlinux

File: compiler/z/codegen/OMRRegisterDependency.cpp [hunk 6]
```diff
@@ -680,167 +680,262 @@ OMR::Z::RegisterDependencyConditions::bookKeepingRegisterUses(TR::Instruction *
    }
 
 uint32_t
-TR_S390RegisterDependencyGroup::genBitVectOfAssignableGPRs(TR::Instruction * currentInstruction, uint32_t          numberOfRegisters,
-   TR::CodeGenerator * cg)
+TR_S390RegisterDependencyGroup::genBitVectOfAssignableGPRs(TR::CodeGenerator * cg,
+                                                           uint32_t numberOfRegisters)
```
[the remark this answers is not in the document]
Comment (NigelYiboYu) [reply compiler/z/codegen/OMRRegisterDependency.cpp]: Changed to `genRegMapOfAssignableGPRs `.

File: compiler/z/codegen/OMRRegisterDependency.cpp [hunk 7]
```diff
@@ -680,167 +680,262 @@ OMR::Z::RegisterDependencyConditions::bookKeepingRegisterUses(TR::Instruction *
...
    {
-   int32_t i;
    uint32_t availRegMap = cg->machine()->genBitVectOfAssignableGPRs();
```
[the remark this answers is not in the document]
Comment (NigelYiboYu) [reply compiler/z/codegen/OMRRegisterDependency.cpp]: Changed to `genRegMapOfAssignableGPRs `.

File: compiler/z/codegen/OMRRegisterDependency.cpp [hunk 8]
```diff
@@ -680,167 +680,262 @@ OMR::Z::RegisterDependencyConditions::bookKeepingRegisterUses(TR::Instruction *
...
-   for (i = 0; i < NUM_S390_GPR; i += 2)
-      {
-      if ((availRegMap >> i) & 0x3 == 0x3)
-         {
-         numAvailPairs++;
-         }
-      }
+/** \brief
+ *     Checks to ensure that we have enough register pairs available to satisfy all register
+ *     dependencies in the current denepency group and aborts the
```
[the remark this answers is not in the document]
Comment (NigelYiboYu) [reply compiler/z/codegen/OMRRegisterDependency.cpp]: Fixed.

File: compiler/z/codegen/OMRRegisterDependency.cpp [hunk 9]
```diff
@@ -680,167 +680,262 @@ OMR::Z::RegisterDependencyConditions::bookKeepingRegisterUses(TR::Instruction *
...
+ *     dependency condition and sets HPR related flags for future register assignment steps.
+ *
+ *  \param cg
+ *     The code generator used to query various machine state.
+ *
+ *  \param currentInstruction
+ *     The instruction on which the register dependency group is attached.
+ *
+ *  \param availableGPRMap
+ *     The bitmask of avilable GRP registers.
```
[the remark this answers is not in the document]
Comment (NigelYiboYu) [reply compiler/z/codegen/OMRRegisterDependency.cpp]: Fixed.

File: compiler/z/codegen/OMRRegisterDependency.cpp [hunk 10]
```diff
@@ -680,167 +680,262 @@ OMR::Z::RegisterDependencyConditions::bookKeepingRegisterUses(TR::Instruction *
...
+         if (realRegI == realRegJ &&
+                 TR::RealRegister::isRealReg(realRegI))
             {
             if (cg->getDebug())
-               cg->getDebug()->printRegisterDependencies(comp->getOutFile(), this, numberOfRegisters);
+               cg->getDebug()->printRegisterDependencies(comp->getOutFile(), this, numOfDependencies);
             TR_ASSERT( 0, "checkDependencyGroup::Real reg dup found in register dependencies\n");
             }
 
+         bool isHighWordSupported = cg->supportsHighWordFacility() && !cg->comp()->getOption(TR_DisableHighWordRA);
```
[the remark this answers is not in the document]
Comment (NigelYiboYu) [reply compiler/z/codegen/OMRRegisterDependency.cpp]: It's moved out of the outermost loop.

File: compiler/z/codegen/OMRRegisterDependency.cpp [hunk 11]
```diff
@@ -680,167 +680,262 @@ OMR::Z::RegisterDependencyConditions::bookKeepingRegisterUses(TR::Instruction *
...
+ *      a bit mask of available GPR map.
+ *      Bit is set to 1 if the GPR is available, and 0 otherwise.
+*/
+uint32_t
+TR_S390RegisterDependencyGroup::checkDependencyGroup(TR::CodeGenerator * cg,
+                                                     TR::Instruction * currentInstruction,
+                                                     uint32_t          numOfDependencies)
+   {
+   // availableGPRMap is a bitmap of all available (not locked) GPRs on the machine
+   // excluding the real registers alreadY specified in the _dependencies.
```
[the remark this answers is not in the document]
Comment (NigelYiboYu) [reply compiler/z/codegen/OMRRegisterDependency.cpp]: Fixed.

**Comment by 0xdaryl** [comment]
genie-omr build zLinux

**Comment by 0xdaryl** [comment]
Z codegen specific change.  It does not require P builds.

**Pull request merged**